### PR TITLE
Switched to custom errors in contracts

### PR DESCRIPTION
**Pull request by aritkulova**
- [ ] Since this PR suggests a **bug fix**, the tests have been added and the coverage is 100%.
- [x] Since this PR introduces a **new feature**, the update has been discussed in an Issue or with the team.
- [ ] This PR is just a **minor change**, like a typo fix.

---

### CODE REVIEW

File: contracts/contracts-registry/pools/AbstractPoolContractsRegistry.sol [hunk 1]
```diff
@@ -82,7 +85,9 @@ abstract contract AbstractPoolContractsRegistry is Initializable, AbstractDepend
     function getProxyBeacon(string memory name_) public view returns (address) {
         address beacon_ = address(_beacons[name_]);
 
-        require(beacon_ != address(0), "PoolContractsRegistry: bad ProxyBeacon");
+        if (beacon_ == address(0)) {
+            revert PoolContractsRegistryBadProxyBeacon();
```
Comment (KyrylR) [inline contracts/contracts-registry/pools/AbstractPoolContractsRegistry.sol:89]: Better something like: does not exist

File: contracts/contracts-registry/AbstractContractsRegistry.sol [hunk 2]
```diff
@@ -215,7 +233,9 @@ abstract contract AbstractContractsRegistry is Initializable {
         string memory name_,
         address contractAddress_
     ) internal virtual {
-        require(contractAddress_ != address(0), "ContractsRegistry: zero address is forbidden");
+        if (contractAddress_ == address(0)) {
+            revert ContractsRegistryZeroAddress();
```
Comment (KyrylR) [inline contracts/contracts-registry/AbstractContractsRegistry.sol:237]: Better to say that contract address is invalid

File: contracts/contracts-registry/AbstractDependant.sol [hunk 3]
```diff
@@ -76,6 +78,8 @@ abstract contract AbstractDependant {
     function _checkInjector() internal view {
         address injector_ = getInjector();
 
-        require(injector_ == address(0) || injector_ == msg.sender, "Dependant: not an injector");
+        if (injector_ != address(0) && injector_ != msg.sender) {
+            revert DependantNotAnInjector();
```
Comment (KyrylR) [inline contracts/contracts-registry/AbstractDependant.sol:82]: Add args please

**Review by KyrylR — CHANGES_REQUESTED** [review]

File: contracts/oracles/UniswapV2Oracle.sol [hunk 4]
```diff
@@ -163,7 +168,9 @@ abstract contract UniswapV2Oracle is Initializable {
      * @param newTimeWindow_ the new time window value in seconds
      */
     function _setTimeWindow(uint256 newTimeWindow_) internal {
-        require(newTimeWindow_ > 0, "UniswapV2Oracle: time window can't be 0");
+        if (newTimeWindow_ == 0) {
+            revert InvalidTimeWindow(0);
+        }
```
Comment (KyrylR) [inline contracts/oracles/UniswapV2Oracle.sol:173]:
```suggestion
        if (newTimeWindow_ == 0) revert InvalidTimeWindow();
```

File: contracts/proxy/transparent/TransparentProxyUpgrader.sol [hunk 5]
```diff
@@ -10,6 +10,8 @@ import {ISolarityTransparentProxy} from "./SolarityTransparentProxy.sol";
  * This is the lightweight helper contract that may be used as a TransparentProxy admin.
  */
 contract TransparentProxyUpgrader is PermanentOwnable {
+    error AddressNotAProxy(address notProxy);
```
Comment (KyrylR) [inline contracts/proxy/transparent/TransparentProxyUpgrader.sol:13]:
```suggestion
    error AddressNotAProxy(address contractAddress);
```

File: contracts/tokens/SBT.sol [hunk 6]
```diff
@@ -156,8 +160,9 @@ abstract contract SBT is ISBT, ERC165Upgradeable {
      * @param tokenId_ the token to mint
      */
     function _mint(address to_, uint256 tokenId_) internal virtual {
-        require(to_ != address(0), "SBT: address(0) receiver");
-        require(!tokenExists(tokenId_), "SBT: token already exists");
+        if (to_ == address(0)) revert Invalidreceiver(address(0));
```
Comment (KyrylR) [inline contracts/tokens/SBT.sol:163]:
```suggestion
        if (to_ == address(0)) revert Invalidreceiver();
```

Also, we could name such error like ReceiverIsZeroAddress

File: contracts/utils/BlockGuard.sol [hunk 7]
```diff
@@ -55,7 +57,7 @@ abstract contract BlockGuard {
      * @param key_ the key of the resource (the caller)
      */
     function _checkBlock(string memory resource_, address key_) internal view {
-        require(_lockedInBlocks[resource_][key_] != _getBlockNumber(), "BlockGuard: locked");
+        if (_lockedInBlocks[resource_][key_] == _getBlockNumber()) revert BlockGuardLocked();
```
Comment (KyrylR) [inline contracts/utils/BlockGuard.sol:60]:
```suggestion
        if (_lockedInBlocks[resource_][key_] == _getBlockNumber()) revert BlockGuardLocked(resource_, key_);
```

**Review by KyrylR — CHANGES_REQUESTED** [review]

File: contracts/diamond/tokens/ERC721/DiamondERC721Storage.sol [hunk 8]
```diff
@@ -39,6 +39,13 @@ abstract contract DiamondERC721Storage is
         mapping(address => mapping(uint256 => uint256)) ownedTokens;
     }
 
+    /**
+     * @dev The owner being `address(0)` indicates a global out of bounds index.
+     */
+    error OutOfBoundsIndex(address owner, uint256 index);
```
[the remark this answers is not in the document]
Comment (aritkulova) [reply contracts/diamond/tokens/ERC721/DiamondERC721Storage.sol:45]: Ok

File: contracts/libs/data-structures/IncrementalMerkleTree.sol [hunk 9]
```diff
@@ -300,11 +303,12 @@ library IncrementalMerkleTree {
         function(bytes32, bytes32) view returns (bytes32) hash2;
     }
 
+    error TreeIsFull();
```
[the remark this answers is not in the document]
Comment (aritkulova) [reply contracts/libs/data-structures/IncrementalMerkleTree.sol:306]: You mean that place of declaration is strange?

File: contracts/diamond/access/ownable/DiamondOwnableStorage.sol [hunk 10]
```diff
@@ -41,6 +43,6 @@ abstract contract DiamondOwnableStorage is InitializableStorage {
      * @notice The function to check if `msg.sender` is the owner
      */
     function _onlyOwner() internal view virtual {
-        require(owner() == msg.sender, "DiamondOwnable: not an owner");
+        if (owner() != msg.sender) revert CallerNotOwner(msg.sender);
```
Comment (KyrylR) [inline contracts/diamond/access/ownable/DiamondOwnableStorage.sol:46]:
```suggestion
        if (owner() != msg.sender) revert CallerNotOwner(msg.sender, owner());
```

File: contracts/diamond/tokens/ERC20/DiamondERC20.sol [hunk 11]
```diff
@@ -139,8 +146,9 @@ contract DiamondERC20 is DiamondERC20Storage {
      * @notice Sets `amount` as the allowance of `spender` over the `owner` s tokens.
      */
     function _approve(address owner_, address spender_, uint256 amount_) internal virtual {
-        require(owner_ != address(0), "ERC20: approve from the zero address");
-        require(spender_ != address(0), "ERC20: approve to the zero address");
+        if (owner_ == address(0)) revert ApproverIsZeroAddress();
+
+        if (spender_ == address(0)) revert SpenderIsZeroAddress();
```
Comment (KyrylR) [inline contracts/diamond/tokens/ERC20/DiamondERC20.sol:151]:
```suggestion
        if (owner_ == address(0)) revert ApproverIsZeroAddress();
        if (spender_ == address(0)) revert SpenderIsZeroAddress();
```

File: contracts/diamond/tokens/ERC721/DiamondERC721Storage.sol [hunk 12]
```diff
@@ -122,7 +126,7 @@ abstract contract DiamondERC721Storage is
      * @notice This function allows you to retrieve the NFT token ID at a given `index` of all the tokens stored by the contract.
      */
     function tokenByIndex(uint256 index_) public view virtual returns (uint256) {
-        require(index_ < totalSupply(), "ERC721Enumerable: global index out of bounds");
+        if (index_ >= totalSupply()) revert GlobalIndexOutOfBounds(index_);
```
Comment (KyrylR) [inline contracts/diamond/tokens/ERC721/DiamondERC721Storage.sol:129]:
```suggestion
        if (index_ >= totalSupply()) revert IndexOutOfBounds(index_);
```

File: contracts/libs/data-structures/AvlTree.sol [hunk 13]
```diff
@@ -420,19 +422,24 @@ library AvlTree {
         function(bytes32, bytes32) view returns (int256) comparator;
     }
 
+    error NodeAlreadyExists(bytes32 key);
```
Comment (KyrylR) [inline contracts/libs/data-structures/AvlTree.sol:425]: Let's define errors by the spec at the top of the file

File: contracts/tokens/SBT.sol [hunk 14]
```diff
@@ -173,7 +178,8 @@ abstract contract SBT is ISBT, ERC165Upgradeable {
      */
     function _burn(uint256 tokenId_) internal virtual {
         address owner_ = _ownerOf(tokenId_);
-        require(owner_ != address(0), "SBT: token doesn't exist");
+
+        if (owner_ == address(0)) revert TokenDoesNotExist(tokenId_);
```
Comment (KyrylR) [inline contracts/tokens/SBT.sol:182]: Typo (do not use `does not`)

**Review by KyrylR — CHANGES_REQUESTED** [review]

File: contracts/libs/data-structures/AvlTree.sol (hunk 13 above)
Comment (aritkulova) [reply contracts/libs/data-structures/AvlTree.sol:425]: okay

**Review by KyrylR — APPROVED** [review]
LGTM

**Pull request merged**